plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2006,7 +2008,13 @@ producer_plugin_impl::push_transaction( const fc::time_point& block_deadline,
       }
    }
 
+   if( trx->is_read_only() ) {
+      chain.set_db_read_only_mode();
+   }
    auto trace = chain.push_transaction( trx, block_deadline, max_trx_time, prev_billed_cpu_time_us, false, sub_bill );
```

Comment:
An exception thrown here leaves chain in read_only mode. Also not sure how this will work when you make it multi-threaded. I would have expected the `set_db_read_only_mode` to be set elsewhere according to a timer.

Reply:
I think it should use a RAII object with a static counter, sets it to read_only when the counter is incremented 0 -> 1, and set to read_write  when the last destructor resets the counter to 0.

Reply:
This is only a temporary place to set the DB read_only_mode. It will be moved to where we toggle single threaded and multi-threaded execution modes when multi-threaded is implemented. But I will add a catch here to prevent DB stuck in read_only mode for now. Thanks.

Reply:
I agree that RAII (`fc::scoped_exit`) would be much cleaner than the try-catch.

Reply:
I did not see Greg's RAII suggestion while replying. Thanks @greg7mdp and @heifner. Will use RAII. That's much better.

Reply:
I think the `scoped_exit` usage is a bit awkward.

Could we add this class in libfc:

```
#pragma once

#include <utility>

namespace fc {

template<class Set, class Unset>
class scoped_set_unset_if {
public:
   scoped_set_unset_if(Set&& set, Unset &&unset, bool do_it = true) :
      unset_(std::forward<Unset>(unset))
      do_it_(do_it)
   {
      if (do_it_)
         std::forward<Set>(set)();
   }

   ~scoped_set_unset_if() {
      if (do_it_)
         std::forward<Unset>(unset_)();
   }

   scoped_set_unset_if(const scoped_set_unset_if&)            = delete;
   scoped_set_unset_if& operator=(const scoped_set_unset_if&) = delete;

private:
   Unset unset_;
   bool  do_it_;
};

} // namespace fc
``` 

and then we could just write:

```
fc::scoped_set_unset_if readonly([]() { chain.set_db_read_only_mode(); },
                                 []() { chain.unset_db_read_only_mode(); },
                                 trx->is_read_only());
```

Reply:
Thanks Greg for the detailed suggestion. `fc::scoped_exit ` is a common pattern in the code base. Instead of introducing more code, I think direct use of it is OK (PS. Setting and unsetting DB read only mode will be moved soon when multi-threading is implemented.

Reply:
OK, but I still think that what I suggested is nicer.